design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+The testing tool also enables users to easily share their test files without any dependency on the execution environment. Therefore, it will be much easier to collaborate and design streaming pipelines with KSQL!
+
+
+## What is in scope
+
+The testing tool will be a self sufficient command line tool where users can pass JSON files describing test settings and configuration and evaluate the generated results.
+
+## What is not in scope
+
+N/A
+
+## Value/Return
+
+The proposed tool provides a new way of testing KSQL statements. It also is self sufficient and does not need any running environment such as Kafka or KSQL cluster.
+
+
+## Public APIS
+
+### Test File
+
+The test file is a JSON file that describes the test settings and test data. The following is an example test file:
+
+```JSON
+{
+ "comments": [
+   "Any comments on this test file!"
+ ],
+ "tests": [
+   {
+     "name": "Sample test",
+     "statements": [
+       "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+       "CREATE STREAM S1 as SELECT name FROM test where id > 100;"
+     ],
+     "inputs": [
+       {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
+       {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
+       {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
+     ],
+     "outputs": [
+       {"topic": "S1", "key": 101, "value": "101"}
+     ]
+   }
+ ]
+}
+```
+
+Each test file can include multiple tests. Each test can have the following attributes (these descriptions are the same as the ones described ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md))):
+
+- _name_ (required): The name of the test case.
+description (optional): A description of what the test case is testing.
+- _format_ (optional): An array of multiple different formats to run the test case as, e.g. AVRO, JSON, DELIMITED.
+- _statements_ (required): The list of statements to execute as this test case.
+- _properties_ (optional): A map of property name to value. Can contain any valid Ksql config. The config is passed to the engine when executing the statements in the test case
+- _topics_ (optional): An array of the topics this test case needs. Allows more information about the topic to be supplied, e.g. an existing Avro schema.
+- _inputs_ (required if `expectedException` not supplied): The set of input messages to be produced to Kafka topic(s).
+- _outputs_ (required if `expectedException` not supplied) The set of output messages expected in the output topic(s).
+- _expectedException_ (required if `inputs` and `outputs` not supplied) The exception that should be thrown when executing the supplied statements.
```

Comment:
note: there is now a `post` node accepted by QTT.  This sets the post conditions.

I would suggest that, as part of this piece of work, we restructure the json documents to more clearly specify pre and post conditions. (We can re-write our own JSON tests to follow the same pattern).

Something like:

```
{
 "comments": [
   "Any comments on this test file!"
 ],
 "tests": [
   {
     "name": "Sample test",
     "comments": "The purpose of this test is...",
     "pre: {
       "statements": [
         "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
         "CREATE STREAM S1 as SELECT name FROM test where id > 100;"
       ]
     },
     "inputs": [
       {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
       {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
       {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
     ],
     "post" {
         "outputs": [
             {"topic": "S1", "key": 101, "value": "101"}
         ],
        "expectedException": {
          "type": "io.confluent.ksql.util.KsqlException",
          "message": "The statement does not define any columns."
         },
        "sources": [
               {"name": "S1", "type": "stream", "keyField": {"name": null}}
        ]
      }
   }
 ]
}
```

The main thing is to move the 'outputs' into the 'post' node as these just another post condition that needs to be checked.

It's debatable if the `statements` node is the thing under test or a precondition. So rather than:

```
{
   "pre": {
        "statements": []
    },
    "inputs": [],
    "post": {
        "outputs:[]
    }
}
```

We could have the inputs as a precondition and make the statements the "bit in the middle under test":

```
{
   "pre": {
        "inputs": []
    },
    "statements": [],
    "post": {
        "outputs:[]
    }
}
```

The benefit of restructuring into `pre` and `post` is it makes it much easier to understand what is a precondition and what is post and generally makes the JSON document more structured, rather than flat.  Also, its more extensible as its possible to have nodes with the same name in both pre and post without a name clash.

Reply:
My first instinct was that having `statements` as a precondition and `inputs` directly under the test feels very wrong since it's the statements are being tested, not the inputs (and it violates intuitive parallel structure in the sense that if `outputs` is under `post`, then it feels natural to have `inputs` under `pre`). Upon further thought, I realize that having `statements` under `pre` and `inputs` directly under the test is inline with how KSQL is used: start up some persistent queries, _then_ produce data to the input topics and observe the outputs produced by the queries. The other proposal of having `inputs` as a precondition and `statements` directly under the test runs counter to this typical usage flow, since it suggests that the input topics are populated first and then the KSQL queries are issued. 

What about a third option with both `inputs` and `statements` as preconditions?
```
{
   "pre": {
        "statements": [],
        "inputs": []
    },
    "post": {
        "outputs:[]
    }
}
```
This option says "given these statements and inputs, we expect the following outputs." We're still moving away from having a flat structure, and users/test-writers can interpret either the statements or inputs as coming first.

Reply:
I like it! :D

Reply:
Maybe just rename to `given` and `then` instead of `pre` and `post` then.

Reply:
I'm okay with these minor changes, though I don't think we need to go terribly far with it. We know that in the end, we want to end up with some REPL-like experience for interactive testing, so that would someday subsume this.